File: Algorithms/Easy/ValidPalindromeII.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def validPalindrome(self, s: str) -> bool:
        l, r = 0, len(s) - 1
        while l < r:
            if s[l] != s[r]:
                return self.isValid(s, l + 1, r) or self.isValid(s, l, r - 1)

            l += 1
            r -= 1

        return True

    def isValid(self, s: str, l: int, r: int) -> bool:
        while l < r:
            if s[l] != s[r]:
                return False

            l += 1
            r -= 1

        return True

File: Algorithms/Easy/ValidPalindromeII.go
package easy

// Time complexity: O(n)
// Space complexity: O(1)
func validPalindrome(s string) bool {
	l, r := 0, len(s)-1
	for l < r {
		if s[l] != s[r] {
			return checkRemain(s, l+1, r) || checkRemain(s, l, r-1)
		}
		l++
		r--
	}

	return true
}

func checkRemain(s string, l int, r int) bool {
	for l < r {
		if s[l] != s[r] {
			return false
		}
		l++
		r--
	}

	return true
}
